harvester_e2e_tests/integrations/test_4_vm_backup_restore.py

```diff
@@ -626,6 +626,210 @@ def test_restore_replace_with_vm_shutdown_command(
+        # take vm snapshot
+        code, data = api_client.vm_snapshots.create(unique_vm_name, vm_snapshot_name)
+        assert 201 == code
+
+        deadline = datetime.now() + timedelta(seconds=wait_timeout)
+        while deadline > datetime.now():
+            code, data = api_client.vm_snapshots.get(vm_snapshot_name)
+            if data.get("status", {}).get("readyToUse"):
+                break
+            print(f"waiting for {vm_snapshot_name} to be ready")
+            sleep(3)
+        else:
+            raise AssertionError(f"timed out waiting for {vm_snapshot_name} to be ready")
+
+        assert 200 == code
+        assert data.get("status", {}).get("readyToUse") is True
+
+        # mess up the existing data
+        with vm_shell_from_host(
+            base_vm_with_data['host_ip'], base_vm_with_data['vm_ip'],
+            base_vm_with_data['ssh_user'], pkey=pri_key
+        ) as sh:
+            out, err = sh.exec_command(f"echo {pub_key!r} > {base_vm_with_data['data']['path']}")
+            assert not err, (out, err)
+            sh.exec_command('sync')
+
+        # Stop the VM then restore existing
+        vm_stopped, (code, data) = vm_checker.wait_stopped(unique_vm_name)
+        assert vm_stopped, (
+            f"Failed to Stop VM({unique_vm_name}) with errors:\n"
+            f"Status({code}): {data}"
+        )
+
+        spec = api_client.backups.RestoreSpec.for_existing(delete_volumes=False)
+        code, data = api_client.backups.restore(unique_vm_name, spec)
+        assert 201 == code, f'Failed to restore backup with current VM replaced, {data}'
+
+        # Check VM Started then get IPs (vm and host)
+        vm_got_ips, (code, data) = vm_checker.wait_ip_addresses(unique_vm_name, ['default'])
+        assert vm_got_ips, (
+            f"Failed to Start VM({unique_vm_name}) with errors:\n"
+            f"Status: {data.get('status')}\n"
+            f"API Status({code}): {data}"
+        )
+        vm_ip = next(iface['ipAddress'] for iface in data['status']['interfaces']
+                     if iface['name'] == 'default')
+        code, data = api_client.hosts.get(data['status']['nodeName'])
+        host_ip = next(addr['address'] for addr in data['status']['addresses']
+                       if addr['type'] == 'InternalIP')
+
+        # Login to the new VM and check data is existing
+        with vm_shell_from_host(host_ip, vm_ip, base_vm_with_data['ssh_user'], pkey=pri_key) as sh:
+            cloud_inited, (out, err) = vm_checker.wait_cloudinit_done(sh)
+            assert cloud_inited, (
+                f"VM {unique_vm_name} Started {wait_timeout} seconds"
+                f", but cloud-init still in {out}"
+            )
+            out, err = sh.exec_command(f"cat {backup_data['path']}")
+
+        assert backup_data['content'] in out, (
+            f"cloud-init writefile failed\n"
+            f"Executed stdout: {out}\n"
+            f"Executed stderr: {err}"
+        )
```

Comment:
Please teardown

Reply:
Thanks for the check and suggestion. 

Indeed after test case `test_with_snapshot_restore_replace_retain_vols` complete execution, the retain volume would exists and all backup, snapshot and vm would be cleaned up by fixture.  

![image](https://github.com/user-attachments/assets/9c0722e0-5df2-4013-b7f6-61f598a13fe8)

I have tried different ways to clean up the retain volume like:
1. Add check clean up the retained volume at the end of the test case
2. Add a new fixture with the name cleanup_retain_volume to cleanup volume right after the base_vm_with_data fixture
3. Add a function inside the test case and use the request.addfinalizer() to call the function 

But the result is it would cause the vm stuck in `Terminating` and all snapshot are also not deleted
  ![image](https://github.com/user-attachments/assets/15296ae4-677a-4bf4-b946-65d1c91106e5)
  ![image](https://github.com/user-attachments/assets/a325ea7b-af6b-408b-b6a9-9bb31492020b)

I also tried with manual operation from UI, when delete the retain volume then delete vm, it works fine. 
While when setting breakpoint and test with api call, the vm would also stuck in `Terminating`

Thus I add the code to check whether the retain volume exists only. 
 
To consider the tradeoff and not cause side effect to other test cases in the file, not sure whether we can keep the retain volume in this test case.

Reply:
Test case aiming for automate UI operations, they should have the same behavior.
However, teardown retained volume is not our main target and has low priority.
If still can not figure out why they behave differently, please comment a `# TODO` or `# FIXME` for come back it later.

Reply:
Thanks for the suggestion. 
I added a `# TODO` to remind the retain volume to think of and come back it later after find the root cause and solution.